src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,862 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "opto/rangeinference.hpp"
+#include "opto/type.hpp"
+#include "utilities/tuple.hpp"
+
+constexpr juint SMALLINT = 3; // a value too insignificant to consider widening
+
+// This represents the result of an iterative calculation
+template <class T>
+class AdjustResult {
+public:
+  bool _progress;             // whether there is progress compared to the last iteration
+  bool _is_result_consistent; // whether the calculation arrives at contradiction
+  T _result;
+
+  static AdjustResult<T> make_empty() {
+    return {true, false, {}};
+  }
+};
+
+// In the canonical form, [lo, hi] intersects with [ulo, uhi] can result in 2
+// cases:
+// - [lo, hi] is the same as [ulo, uhi], lo and hi are both >= 0 or both < 0.
+// - [lo, hi] is not the same as [ulo, uhi], which results in the intersections
+// being [lo, uhi] and [ulo, hi], lo and uhi are < 0 while ulo and hi are >= 0.
+// This class deals with each interval with both bounds being >= 0 or < 0 in
+// the signed domain.
+template <class U>
+class SimpleCanonicalResult {
+  static_assert(std::is_unsigned<U>::value, "bit info should be unsigned");
+public:
+  bool _present; // whether this is an empty set
+  RangeInt<U> _bounds;
+  KnownBits<U> _bits;
+
+  static SimpleCanonicalResult<U> make_empty() {
+    return {false, {}, {}};
+  }
+};
+
+// Find the minimum value that is not less than lo and satisfies bits.
+// Here, we view a number in binary as a bit string. As a result,  the first
+// bit refers to the highest bit (the MSB), the last bit refers to the lowest
+// bit (the LSB), a bit comes before (being higher than) another if it is more
+// significant, and a bit comes after (being lower than) another if it is less
+// significant.
+template <class U>
+static U adjust_lo(U lo, const KnownBits<U>& bits) {
+  constexpr size_t W = sizeof(U) * 8;
+  // Violation of lo with respects to bits
+  // E.g: lo    = 1100
+  //      zeros = 0100
+  //      ones  = 1001
+  // zero_violation = 0100, i.e the second bit should be zero, but it is 1 in
+  // lo. Similarly, one_violation = 0001, i.e the last bit should be one, but
+  // it is 0 in lo. These make lo not satisfy the bit constraints, which
+  // results in us having to find the smallest value that satisfies bits
+  U zero_violation = lo & bits._zeros;
+  U one_violation = ~lo & bits._ones;
+  if (zero_violation == one_violation) {
+    // This means lo does not violate bits, it is the result
+    assert(zero_violation == 0, "");
+    return lo;
+  }
+
+  /*
```

Comment:
Everywhere else you use `//`, and that seems to be generally our style, so I'd keep it consistent ;)

Reply:
From https://github.com/openjdk/jdk/pull/9947, I learnt that for such a long paragraph with dense mathematics, reducing `//` noise would make reading it easier.